libbeat/processors/add_cloud_metadata/add_cloud_metadata.go

```diff
@@ -209,30 +248,137 @@ func fetchMetadata(doURL, ec2URL, gceURL string, timeout time.Duration) *result
+func makeCommomJSONPicker(provider string) pick {
+	return func(all []byte, res *result) error {
+		dec := json.NewDecoder(bytes.NewReader(all))
+		dec.UseNumber()
+		err := dec.Decode(&res.metadata)
+		if err != nil {
+			err = errors.Wrapf(err, "failed to unmarshal %v JSON of '%v'", provider, string(all))
+			return err
+		}
+		return nil
+	}
+}
 
-	var (
-		doURL  = "http://" + config.MetadataHostAndPort + doMetadataURI
-		ec2URL = "http://" + config.MetadataHostAndPort + ec2InstanceIdentityURI
-		gceURL = "http://" + config.MetadataHostAndPort + gceMetadataURI
-	)
+// newMetadataFetcher return metadataFetcher with one pass json picker
+func newMetadataFetcher(
+	c common.Config,
+	timeout time.Duration,
+	provider string,
+	headers map[string]string,
+	host string,
+	conv schemaConv,
+	uris []string,
+) (*metadataFetcher, error) {
+	urls, err := getMetadataURL(c, host, timeout, uris)
+	if err != nil {
+		return nil, err
+	}
+	picker := map[string]pick{urls[0]: makeCommomJSONPicker(provider)}
+	fetcher := &metadataFetcher{provider, headers, picker, conv}
+	return fetcher, nil
+}
 
-	result := fetchMetadata(doURL, ec2URL, gceURL, config.Timeout)
+func newDoMetadataFetcher(c common.Config, timeout time.Duration) (*metadataFetcher, error) {
+	fetcher, err := newMetadataFetcher(c, timeout, "digitalocean", nil, metadataHost, doSchema, []string{
+		doMetadataURI,
+	})
+	return fetcher, err
+}
+
+func newEc2MetadataFetcher(c common.Config, timeout time.Duration) (*metadataFetcher, error) {
+	fetcher, err := newMetadataFetcher(c, timeout, "ec2", nil, metadataHost, ec2Schema, []string{
+		ec2InstanceIdentityURI,
+	})
+	return fetcher, err
+}
+
+func newGceMetadataFetcher(c common.Config, timeout time.Duration) (*metadataFetcher, error) {
+	fetcher, err := newMetadataFetcher(c, timeout, "gce", gceHeaders, metadataHost, gceSchema, []string{
+		gceMetadataURI,
+	})
+	return fetcher, err
+}
+
+// newQcloudMetadataFetcher return the concrete metadata fetcher for qcloud provider
+// which requires more than one way to assemble the metadata
+func newQcloudMetadataFetcher(c common.Config, timeout time.Duration) (*metadataFetcher, error) {
+	urls, err := getMetadataURL(c, qcloudMetadataHost, timeout, []string{
+		qcloudMetadataInstanceIdURI,
+		qcloudMetadataRegionURI,
+		qcloudMetadataZoneURI,
+	})
+	if err != nil {
+		return nil, err
+	}
+	picker := map[string]pick{
+		urls[0]: func(all []byte, result *result) error {
+			result.metadata["instance_id"] = string(all)
+			return nil
+		},
+		urls[1]: func(all []byte, result *result) error {
+			result.metadata["region"] = string(all)
+			return nil
+		},
+		urls[2]: func(all []byte, result *result) error {
+			result.metadata["zone"] = string(all)
+			return nil
+		},
+	}
+	fetcher := &metadataFetcher{"qcloud", nil, picker, qcloudSchema}
+	return fetcher, nil
+}
+
+func newCloudMetadata(c common.Config) (processors.Processor, error) {
+	timeout := 3 * time.Second
```

Comment:
yep, it's hardcoded here, already move timeout configuration pack out of the getMetadataURLs and put it directly in newCloudMetadata.
The pool config is packed 5 times here 🤒, but I think it is acceptable here since it only run once in the startup phase.